pkg/controller/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,318 @@
+
+	return err
+}
+
+func (c *CacheManager) RemoveObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	if _, err := c.opa.RemoveData(ctx, instance); err != nil {
+		return err
+	}
+
+	// only delete from metrics map if the data removal was succcesful
+	c.syncMetricsCache.DeleteObject(syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName()))
+	c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+
+	return nil
+}
+
+func (c *CacheManager) wipeData(ctx context.Context) error {
+	if _, err := c.opa.RemoveData(ctx, target.WipeData()); err != nil {
+		return err
+	}
+
+	// reset sync cache before sending the metric
+	c.syncMetricsCache.ResetCache()
+	c.syncMetricsCache.ReportSync()
+
+	return nil
+}
+
+func (c *CacheManager) ReportSyncMetrics() {
+	c.syncMetricsCache.ReportSync()
+}
+
+func (c *CacheManager) listAndSyncDataForGVK(ctx context.Context, gvk schema.GroupVersionKind) error {
+	u := &unstructured.UnstructuredList{}
+	u.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind + "List",
+	})
+
+	err := c.reader.List(ctx, u)
+	if err != nil {
+		return fmt.Errorf("replaying data for %+v: %w", gvk, err)
+	}
+
+	defer c.ReportSyncMetrics()
+
+	for i := range u.Items {
+		if err := c.AddObject(ctx, &u.Items[i]); err != nil {
+			return fmt.Errorf("adding data for %+v: %w", gvk, err)
+		}
+	}
+
+	return nil
+}
+
+func (c *CacheManager) updateDatastore(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-c.replayTicker.C:
+			// snapshot the current spec so we can make a step upgrade
+			// to the contests of the opa cache.
+			c.mu.RLock()
+			currentGVKsInAgg := watch.NewSet()
+			currentGVKsInAgg.Add(c.gvkAggregator.ListAllGVKs()...)
+			excluderChanged := c.excluderChanged
+			c.mu.RUnlock()
+
+			c.makeUpdatesForSpecInTime(ctx, currentGVKsInAgg, excluderChanged)
+		}
+	}
+}
+
+// listAndSyncData returns a set of gvks that were successfully listed and synced.
+func (c *CacheManager) listAndSyncData(ctx context.Context, gvks []schema.GroupVersionKind, reader client.Reader) *watch.Set {
+	gvksSuccessfullySynced := watch.NewSet()
+	for _, gvk := range gvks {
+		err := c.listAndSyncDataForGVK(ctx, gvk)
+		if err != nil {
+			log.Error(err, "internal: error syncing gvks cache data")
+			// we don't remove this gvk as we will try to re-add it later
+			// we also don't return on this error to be able to list and sync
+			// other gvks in order to protect against a bad gvk.
+		} else {
+			gvksSuccessfullySynced.Add(gvk)
+		}
+	}
+	return gvksSuccessfullySynced
+}
+
+// makeUpdatesForSpecInTime performs a conditional wipe followed by a replay if necessary as
+// given by the current spec (currentGVKsInAgg, excluderChanged) at the time of the call.
+func (c *CacheManager) makeUpdatesForSpecInTime(ctx context.Context, currentGVKsInAgg *watch.Set, excluderChanged bool) {
```

Comment:
If there have been other changes mid-replay, do we want to make them interruptable?

Note that whether we are relying on this do add net-new GVKs (as opposed to just handling wipe/replay) may change the answer.

Reply:
as of now, changes mid replay are picked up on the next tick and the current func is by design not interrupt-able

I think if we move to a model where we don't "snapshot" the spec that can cause a lot of race-y behavior && increase complexity.

Reply:
ASAP syncing is a UX issue. Users will add a sync and expect to see the constraint working. Delays cause frustration/bug reports.

Per original requirements, and to preserve existing behavior net new GVKs need to be synced ASAP